docs/RFCS/stateless_replica_relocation.md

```diff
@@ -0,0 +1,232 @@
+- Feature Name: Stateless Replica Relocation
+- Status: draft
+- Start Date: 2015-08-19
+- RFC PR:
+- Cockroach Issue: #620
+
+# Summary
+A relocation is, conceptually, the transfer of a single replica from one store to
+another. However, the implementation is necessarily the combination of two operations:
+
+1. Creating a new replica for a range on a new store.
+2. Removing a replica of the same range from a different store.
+
+For example, by creating a replica on store Y and then removing a replica from
+store X, you have in effect moved the replica from X to Y.
+
+This PR is suggesting an overall architectural design goal: that the decision to
+make any individual operation (either a create or a remove) should be **stateless**.
+In other words, the second operation in a replica relocation should not depend on a
+specific invocation of the first operation.
+
+# Motivation
+For an assortment of reasons, Cockroach will often need to relocate the replicas
+of a range. Most immediately, this is needed for repair (when a store dies and
+its replicas are no longer usable) and rebalance (relocating replicas on
+overloaded stores to stores with excess capacity).
+
+In order to retain the integrity of a range's raft consensus group, a relocation
+must be expressed as a combination of two operations:
+
+1. Creating a new replica for a range on a new store.
+2. Removing a replica of the same range from a different store.
+
+These operations can happen in either order as long as quorum is maintained in
+the range's raft group after each individual operation, although one ordering may be
+preferred over another.
```

Comment:
For rebalancing when all nodes are basically healthy, add-then-remove is definitely better (to minimize the period of reduced availability, especially once we have non-voting replicas or preemptive snapshots). For repair, it only makes a difference to the extent that the "down" node could be a transient problem that recovers during the replication process.

This design, however, seems on the surface to encourage remove-then-add (at least in the repair case), since repair decisions are made when we notice a node that should be removed. In order to repair with add-then-remove, we'd need to notice that a node is down, then essentially forget about that fact and add a new replica instead (with the expectation that we will rediscover the down node's state when we consider the now-overreplicated range). This is a little odd, but should work.